simulator/infra/replacement/cache_replacement.cpp

```diff
@@ -68,10 +69,131 @@ std::size_t LRUCacheInfo::update()
     return lru_elem;
 }
 
+//////////////////////////////////////////////////////////////////
+
+struct node
+{
+    node(int new_value = 0, std::size_t new_pointer = 0)
+    : value( new_value), pointer( new_pointer) {}
+
+    int value;
+    std::size_t pointer;
+};
+
+bool operator == (const node &lhs, const node &rhs) { return lhs.value == rhs.value; }
+bool operator< (const node &lhs, const node &rhs) { return lhs.value < rhs.value; }
+
+class Pseudo_LRUCacheInfo : public CacheReplacementInterface
+{
+    public:
+        explicit Pseudo_LRUCacheInfo( std::size_t ways);
+        void touch( std::size_t way) override;
+        void set_to_erase( std::size_t ) override;
+        std::size_t update() override;
+        std::size_t get_ways() const override { return ways; }
+
+    private:
+        enum flags { Left = 0, Right = 1};
+
+        core::tree<node> lru_tree;
+        void construct_tree( core::tree<node>::iterator node_it, std::size_t max_depth);
+        void construct_leaf_layer();
+        std::size_t calculate_depth() const;
+        std::size_t which_sibling( core::tree<node>::iterator node_it);
+        static int reverse_flag( int flag);
+
+        const std::size_t ways;
+        int leaf_iterator = 0;
+        int node_iterator = INT8_MIN; //make shure it is less than 0
+};
+
+Pseudo_LRUCacheInfo::Pseudo_LRUCacheInfo( std::size_t ways)
+    : ways( ways)
+{
+    lru_tree.data( node(node_iterator));
+    construct_tree( lru_tree.get_tree_iterator(), calculate_depth());
+}
+
+std::size_t Pseudo_LRUCacheInfo::calculate_depth() const
+{
+    std::size_t i = 0;
+        for ( std::size_t numerator = ways; numerator != 1; numerator = numerator/2)
+        {
+            if (numerator%2 != 0)
+                throw CacheReplacementException("Number of ways must be the power of 2!");
```

Comment:
We have functions to do both operations:

https://github.com/MIPT-ILab/mipt-mips/blob/7884c29786b375528f939ed85481e36eb96f93ee/simulator/infra/macro.h#L21

https://github.com/MIPT-ILab/mipt-mips/blob/7884c29786b375528f939ed85481e36eb96f93ee/simulator/infra/macro.h#L115